File: packages/sdk-go/mfa.go
package driftstack

import (
	"context"
	"time"
)

// MfaResource handles /v1/account/mfa/* endpoints (V-353b / V-448).
//
// Pairs with `client.Auth.MfaChallenge` (login MFA exchange) and
// `client.Auth.MfaStepUp` (V-353e step-up gate). MFA enrollment is
// per-account, never per-team-context — these endpoints don't honor
// the X-Driftstack-Account header.
type MfaResource struct {
	client *Client
}

// MfaStatus — V-353b enrollment state.
type MfaStatus struct {
	Enrolled            bool       `json:"enrolled"`
	EnrolledAt          *time.Time `json:"enrolled_at"`
	LastUsedAt          *time.Time `json:"last_used_at"`
	UnusedRecoveryCodes int        `json:"unused_recovery_codes"`
}

// MfaEnrollResponse — first half of TOTP enrollment. Customer scans
// `OtpauthURI` with their authenticator app, then calls Verify with
// the first 6-digit code. SecretBase32 is shown ONCE for manual
// entry; the server stores it encrypted at rest.
type MfaEnrollResponse struct {
	OtpauthURI    string `json:"otpauth_uri"`
	SecretBase32  string `json:"secret_base32"`
	Algorithm     string `json:"algorithm"`      // "SHA1"
	Digits        int    `json:"digits"`         // 6
	PeriodSeconds int    `json:"period_seconds"` // 30
}

// MfaVerifyRequest — first 6-digit TOTP code from the customer's app.
type MfaVerifyRequest struct {
	Code string `json:"code"`
}

// MfaVerifyResponse — 10 single-use recovery codes. Shown ONCE.
type MfaVerifyResponse struct {
	RecoveryCodes []string `json:"recovery_codes"`
}

// MfaDisableRequest — literal "disable-mfa" confirmation phrase.
type MfaDisableRequest struct {
	Confirm string `json:"confirm"` // "disable-mfa"
}

// Status — read MFA enrollment state for the calling account.
func (r *MfaResource) Status(ctx context.Context) (*MfaStatus, error) {
	var out MfaStatus
	if err := r.client.do(ctx, requestOptions{method: "GET", path: "/v1/account/mfa", out: &out}); err != nil {
		return nil, err
	}
	return &out, nil
}

// Enroll — start TOTP enrollment. Customer scans the otpauth URI
// from the response in their authenticator app, then calls Verify.
func (r *MfaResource) Enroll(ctx context.Context) (*MfaEnrollResponse, error) {
	var out MfaEnrollResponse
	if err := r.client.do(ctx, requestOptions{
		method: "POST",
		path:   "/v1/account/mfa/enroll",
		body:   struct{}{},
		out:    &out,
	}); err != nil {
		return nil, err
	}
	return &out, nil
}

// Verify — confirm enrollment with first 6-digit code. Returns 10
// single-use recovery codes (shown ONCE).
func (r *MfaResource) Verify(ctx context.Context, body *MfaVerifyRequest) (*MfaVerifyResponse, error) {
	var out MfaVerifyResponse
	if err := r.client.do(ctx, requestOptions{
		method: "POST",
		path:   "/v1/account/mfa/verify",
		body:   body,
		out:    &out,
	}); err != nil {
		return nil, err
	}
	return &out, nil
}

// Disable — disable MFA. Requires fresh MFA proof per V-353e step-up
// gate. Customer should call MfaStepUp(ctx, ...) first if the 15-min
// window is stale. Recovery codes are invalidated.
func (r *MfaResource) Disable(ctx context.Context, body *MfaDisableRequest) error {
	return r.client.do(ctx, requestOptions{
		method: "DELETE",
		path:   "/v1/account/mfa",
		body:   body,
	})
}

// RegenerateRecoveryCodes — mint 10 fresh recovery codes; old codes
// invalidated. Shown ONCE.
func (r *MfaResource) RegenerateRecoveryCodes(ctx context.Context) (*MfaVerifyResponse, error) {
	var out MfaVerifyResponse
	if err := r.client.do(ctx, requestOptions{
		method: "POST",
		path:   "/v1/account/mfa/recovery-codes/regenerate",
		body:   struct{}{},
		out:    &out,
	}); err != nil {
		return nil, err
	}
	return &out, nil
}

File: packages/sdk-typescript/src/resources/mfa.ts
// MfaResource — typed methods for /v1/account/mfa/*.
//
// Enrollment management (status / enroll / verify / disable / regenerate
// recovery codes). Uses the calling web-session bearer; the
// X-Driftstack-Account team-RBAC header is not honored — MFA is per-
// account, not per-team-context.
//
// Pairs with `client.auth.mfaChallenge` (login MFA exchange) +
// `client.auth.mfaStepUp` (the step-up gate).

import type { HttpClient } from '../http.js';

export interface MfaStatusResponse {
  enrolled: boolean;
  enrolled_at: string | null;
  last_used_at: string | null;
  unused_recovery_codes: number;
}

export interface MfaEnrollResponse {
  /** otpauth:// URI for QR-code rendering in an authenticator app. */
  otpauth_uri: string;
  /** Plaintext base32-encoded TOTP secret for manual entry. */
  secret_base32: string;
  algorithm: 'SHA1';
  digits: 6;
  period_seconds: 30;
}

export interface MfaVerifyRequest {
  /** First 6-digit TOTP code from the customer's authenticator app. */
  code: string;
  /**
   * The account's current password. Required when the account has one; an
   * account without a password (created by Google or GitHub sign-in) leaves it
   * out and needs a sign-in from the last ten minutes instead.
   */
  current_password?: string;
}

export interface MfaVerifyResponse {
  /** 10 single-use recovery codes; shown ONCE. */
  recovery_codes: string[];
}

export interface MfaDisableRequest {
  /** Literal 'disable-mfa' confirmation phrase. */
  confirm: 'disable-mfa';
}

export class MfaResource {
  constructor(private readonly http: HttpClient) {}

  /** Read MFA enrollment state for the calling account. */
  status(): Promise<MfaStatusResponse> {
    return this.http.request<MfaStatusResponse>({
      method: 'GET',
      path: '/v1/account/mfa',
    });
  }

  /**
   * Start TOTP enrollment. Customer scans `otpauth_uri` with their
   * authenticator app, then calls `verify(...)` with the first
   * 6-digit code. Server stores the secret encrypted at rest;
   * plaintext is shown ONCE here.
   */
  enroll(): Promise<MfaEnrollResponse> {
    return this.http.request<MfaEnrollResponse>({
      method: 'POST',
      path: '/v1/account/mfa/enroll',
      body: {},
    });
  }

  /** Confirm enrollment with the first code. Returns 10 recovery codes. */
  verify(body: MfaVerifyRequest): Promise<MfaVerifyResponse> {
    return this.http.request<MfaVerifyResponse>({
      method: 'POST',
      path: '/v1/account/mfa/verify',
      body,
    });
  }

  /**
   * Disable MFA. Requires fresh MFA proof per the step-up gate
   * (15-minute freshness window) — call `client.auth.mfaStepUp(...)`
   * first if the gate is stale. Recovery codes are invalidated.
   */
  disable(body: MfaDisableRequest): Promise<void> {
    return this.http.request<void>({
      method: 'DELETE',
      path: '/v1/account/mfa',
      body,
    });
  }

  /** Mint 10 fresh recovery codes. Old codes invalidated; shown ONCE. */
  regenerateRecoveryCodes(): Promise<MfaVerifyResponse> {
    return this.http.request<MfaVerifyResponse>({
      method: 'POST',
      path: '/v1/account/mfa/recovery-codes/regenerate',
      body: {},
    });
  }
}
